### PR TITLE
[Model] Add ResTransformer as ABINet Backbone

### PR DESCRIPTION
**Pull request by gaotongxiao**
To be reviewed after #521

### CODE REVIEW

File: mmocr/models/textrecog/backbones/resnet_abi.py [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+                 base_channels=3,
+                 arch_settings=[3, 4, 6, 6, 3],
+                 channels=[32, 32, 64, 128, 256, 512],
+                 strides=[1, 2, 1, 2, 1, 1],
+                 out_indices=None,
+                 last_stage_pool=False,
+                 init_cfg=[
+                     dict(type='Xavier', layer='Conv2d'),
+                     dict(type='Constant', val=1, layer='BatchNorm2d')
+                 ]):
```
Comment (Harold-lkk) [inline mmocr/models/textrecog/backbones/resnet_abi.py:33]: Recommend to keep pace with mmcv/mmdet'ResNet
```suggestion
                 base_channels=3,
                 stem_channels=None,
                 base_channel=32,
                 num_stages=5,
                 arch_settings=[3, 4, 6, 6, 3],
                 channels=[32, 64, 128, 256, 512],
                 strides=[1, 2, 1, 2, 1, 1],
                 out_indices=None,
                 last_stage_pool=False,
                 init_cfg=[
                     dict(type='Xavier', layer='Conv2d'),
                     dict(type='Constant', val=1, layer='BatchNorm2d')
                 ]):  
```
it's weird that channels parameter has different dim with other

File: mmocr/models/textrecog/backbones/resnet_abi.py [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+        self.conv1 = nn.Conv2d(
+            base_channels,
+            channels[0],
+            kernel_size=3,
+            stride=strides[0],
+            padding=1)
+        self.bn1 = nn.BatchNorm2d(channels[0])
+        self.relu1 = nn.ReLU(inplace=True)
+
+        self.layers = Sequential(*[
```
Comment (Harold-lkk) [inline mmocr/models/textrecog/backbones/resnet_abi.py:56]: Recommend to keep pace with mmcv/mmdet'ResNet

Also channels[0] has no clear meaning

File: mmocr/models/textrecog/backbones/resnet_abi.py [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+            self._make_layer(channels[i], channels[i + 1], arch_settings[i],
+                             strides[i + 1])
+            for i in range(len(arch_settings))
+        ])
+
+    def _make_layer(self, input_channels, output_channels, blocks, stride=1):
```
Comment (Harold-lkk) [inline mmocr/models/textrecog/backbones/resnet_abi.py:62]: Can be replaced by function?
https://github.com/open-mmlab/mmcv/blob/7207397de3c2ffaef878628a55fa2ab37ac0bbc7/mmcv/cnn/resnet.py#L143
Recommend use block=BasicBlock as parameters instead of fixed

File: mmocr/models/textrecog/encoders/restransformer.py [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+                 d_model=512,
+                 d_inner=2048,
+                 dropout=0.1,
+                 max_len=8 * 32,
+                 init_cfg=None):
+        super().__init__(init_cfg=init_cfg)
+
+        assert d_model % n_head == 0, 'd_model must be divisible by n_head'
+
+        self.resnet = ResNetABI()
```
Comment (Harold-lkk) [inline mmocr/models/textrecog/encoders/restransformer.py:37]: self.cnn_backbone = build_backbone(cnn_backbone)

File: mmocr/models/textrecog/encoders/restransformer.py [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+    """
+
+    def __init__(self,
+                 n_layers=2,
+                 n_head=8,
+                 d_model=512,
+                 d_inner=2048,
+                 dropout=0.1,
+                 max_len=8 * 32,
+                 init_cfg=None):
```
Comment (Harold-lkk) [inline mmocr/models/textrecog/encoders/restransformer.py:32]: add a parameter for config ResNet:
    cnn_backbone = dict(type=ReaNetABI, ...)
otherwise, it cannot change ResNetABI parameter

File: mmocr/models/textrecog/backbones/resnet_abi.py (hunk 3 above)
Comment (Harold-lkk) [reply mmocr/models/textrecog/backbones/resnet_abi.py:65]:
```suggestion
    def _make_res_layer(block,
                        inplanes,
                        planes,
                        blocks,
                        stride=1,
                        dilation=1
```
keep the same API with upstream(mmcv/mmdet)

File: mmocr/models/textrecog/backbones/resnet_abi.py [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+
+        self.layers = [
+            self._make_layer(stem_channels, stem_channels, arch_settings[0],
+                             strides[0])
+        ]
+        for i in range(1, len(arch_settings)):
+            self.layers.append(
+                self._make_layer(stem_channels * 2**(i - 1),
+                                 stem_channels * 2**i, arch_settings[i],
+                                 strides[i]))
```
Comment (Harold-lkk) [inline mmocr/models/textrecog/backbones/resnet_abi.py:62]:
```suggestion
        for i, num_blocks in enumerate(self.stage_blocks)::
            stride = strides[i]
            dilation = dilations[i]
            planes = base_channels * 2**i
            res_layer = self.make_res_layer(
                block=self.block,
                inplanes=self.inplanes,
                planes=planes,
                num_blocks=num_blocks,
                stride=stride,
                dilation=dilation,
                )
            self.inplanes = planes * block.expansion
```
Could be replaced by https://github.com/open-mmlab/mmdetection/blob/f08548bfd6d394a82566022709b5ce9e6b0a855e/mmdet/models/backbones/resnet.py#L468 or https://github.com/open-mmlab/mmcv/blob/9b227ecfc07d595cab6680e4e86977dd6bf7800a/mmcv/cnn/resnet.py#L244
we should try to implement the code same as upstream, which can only need users to master　one code.

Comment (gaotongxiao) [reply mmocr/models/textrecog/backbones/resnet_abi.py]: Adopting this snippet unnecessarily increases the complexity as it involves many redundant variables and also cannot be directly applied to our case. Sticking to the current implementation should be fine.

File: mmocr/models/textrecog/backbones/resnet_abi.py [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+                planes=planes,
+                blocks=num_blocks,
+                stride=stride)
+            self.inplanes = planes * self.block.expansion
+            planes *= 2
+            layer_name = f'layer{i + 1}'
+            self.add_module(layer_name, res_layer)
+            self.res_layers.append(layer_name)
+
+    def _make_layer(self, block, inplanes, planes, blocks, stride=1):
```
Comment (Harold-lkk) [inline mmocr/models/textrecog/backbones/resnet_abi.py:72]:
```suggestion
    def _make_res_layer(self, block, inplanes, planes, blocks, stride=1):
```

File: mmocr/models/textrecog/encoders/restransformer.py [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+            ffn_cfgs=dict(
+                type='FFN',
+                embed_dims=d_model,
+                feedforward_channels=d_inner,
+                ffn_drop=dropout,
+            ),
+            norm_cfg=dict(type='LN'),
+        )
+        self.transformer = ModuleList(
+            [copy.deepcopy(encoder_layer) for _ in range(n_layers)])
```
Comment (Harold-lkk) [inline mmocr/models/textrecog/encoders/restransformer.py:61]: remove the ResTransformer and only use the ResNetABI for backbone
This transform should be encode of vision branch
```suggestion
    @ENCODE.register_module()
    class ABINetVisionEncode(BaseEncode)
        self.pos_encoder = PositionalEncoding(d_model, n_position=max_len)
        encoder_layer = BaseTransformerLayer(
            operation_order=('self_attn', 'norm', 'ffn', 'norm'),
            attn_cfgs=dict(
                type='MultiheadAttention',
                embed_dims=d_model,
                num_heads=n_head,
                attn_drop=dropout,
                dropout_layer=dict(type='Dropout', drop_prob=dropout),
            ),
            ffn_cfgs=dict(
                type='FFN',
                embed_dims=d_model,
                feedforward_channels=d_inner,
                ffn_drop=dropout,
            ),
            norm_cfg=dict(type='LN'),
        )
        self.transformer = ModuleList(
            [copy.deepcopy(encoder_layer) for _ in range(n_layers)])
```
